src/main/java/us/abstracta/wiresham/ConnectionFlowDriver.java

```diff
@@ -33,42 +38,92 @@ public void run() {
+        processParallelSteps((ParallelPacketStep) step, previousPort, connectionProvider);
+        continue;
+      }
+      if (step.getPort() != null && step.getPort() != previousPort) {
+        LOG.info("changing to connections on port {}", step.getPort());
+        previousPort = step.getPort();
+      }
+      FlowConnection flowConnection = connectionProvider.get(previousPort);
+      step.process(flowConnection);
+    }
+  }
+
+  private static void processParallelSteps(ParallelPacketStep step, int previousPort,
+      FlowConnectionProvider connectionProvider) {
+    List<List<PacketStep>> parallelSteps = step.getParallelSteps();
+    ExecutorService parallelStepsExecutor = Executors.newFixedThreadPool(parallelSteps.size());
+    List<Future<?>> parallelFutures = new ArrayList<>();
+    for (List<PacketStep> parallelStep : parallelSteps) {
+      Queue<PacketStep> steps = new LinkedList<>(parallelStep);
+      parallelFutures.add(parallelStepsExecutor.submit(() -> {
+        try {
+          processSteps(steps, previousPort, connectionProvider);
+        } catch (ExecutionException | InterruptedException | IOException e) {
+          handleProcessStepExceptions(e);
+        }
+      }));
+    }
+    joinParallelExecutions(parallelStepsExecutor, parallelFutures);
+  }
+
+  private static void handleProcessStepExceptions(Exception e) {
+    if (e.getClass().equals(IOException.class)) {
       if (e.getMessage().contains("Socket is closed")) {
         LOG.trace("Received expected exception when server socket has been closed", e);
       } else {
         LOG.error("Problem while processing requests from client. Closing connection.", e);
       }
-    } catch (InterruptedException e) {
+    } else if (e.getClass().equals(InterruptedException.class)) {
       LOG.trace("The thread has been interrupted", e);
       Thread.currentThread().interrupt();
-    } catch (ExecutionException e) {
+    } else if (e.getClass().equals(ExecutionException.class)) {
       LOG.error("Problem while waiting for socket to be created", e);
-    } finally {
-      try {
-        closeFlowConnections();
-      } catch (IOException e) {
-        LOG.error("Problem while releasing sockets", e);
+    } else if (e.getClass().equals(ConnectionClosedException.class)) {
+      LOG.info("Connection closed by client while waiting for client packet");
+      ConnectionClosedException ex = (ConnectionClosedException) e;
+      if (ex.getDiscardedPacket().getBytes().length > 0) {
+        LOG.debug("Discarding client packet {}", ex.getDiscardedPacket(), e);
       }
-      MDC.clear();
     }
   }
 
+  private static void joinParallelExecutions(ExecutorService parallelStepsExecutor,
+      List<Future<?>> parallelFutures) {
+    parallelStepsExecutor.shutdown();
+    parallelFutures.forEach(future -> {
+      try {
+        future.get();
+      } catch (InterruptedException e) {
+        LOG.error("Parallel steps where interrupted", e);
```

Comment:
Have you tested this? Is it ok to throw an error? I mean, user sends ctrl+c and log error? maybe a warning?, but something important is the amount of logging you generate for one event as well.